File: cpp/nc-150/1D-DP/coin_change.cpp
#include <vector>

using namespace std;

/*
 * Author: Valerie Doan
 * Date: October 16, 2024
 * Problem: Coin Change
 * Level: Medium
 * Neetcode Course: Neetcode 150
 * Neetcode Topic: 1-D DP
 */

int coinChange(vector<int> &coins, int amount)
{
    int table[amount + 1];
    fill_n(table, amount + 1, -1);
    table[0] = 0;

    for (int i = 1; i <= amount; i++)
    {
        for (int coin : coins)
        {
            if (i - coin >= 0 && table[i - coin] != -1)
            {
                if (table[i] == -1)
                {
                    table[i] = 1 + table[i - coin];
                }
                else
                {
                    table[i] = min(table[i], 1 + table[i - coin]);
                }
            }
        }
    }

    return table[amount];
}

File: python/nc-150/1D-DP/coin_change.py
"""
Author: Valerie Doan
Date: November 11, 2024
Problem: Coin Change
Level: Medium
Neetcode Course: Neetcode 150
Neetcode Topic: 1-D DP
"""


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [-1 for i in range(amount + 1)]
        dp[0] = 0

        for amt in range(len(dp)):
            for coin in coins:
                if amt - coin < 0 or dp[amt - coin] == -1:
                    continue

                if dp[amt] == -1:
                    dp[amt] = 1 + dp[amt - coin]

                dp[amt] = min(dp[amt], 1 + dp[amt - coin])

        return dp[amount]
